Lib/test/test_concurrent_futures.py

```diff
@@ -746,7 +746,22 @@ def record_finished(n):
 
         self.executor.map(record_finished, range(10))
         self.executor.shutdown(wait=True)
-        self.assertCountEqual(finished, range(10))
+        # No guarantees on how many tasks dispatched,
+        # but at least one should have been dispatched
+        self.assertGreater(len(finished), 0)
```

Comment:
I think this change breaks compatibility. The doc for `shutdown` says:

> If wait is True then this method will not return until all the pending futures are done executing and the resources associated with the executor have been freed.

So all futures should have executed, instead of being cancelled.

Reply:
At the time I wrote it, it didn't conflict with the documentation precisely; the original documentation said that map was "Equivalent to map(func, *iterables) except func is executed asynchronously and several calls to func may be made concurrently.", but doesn't guarantee that any actual futures exist (it's implemented in terms of submit and futures, but doesn't actually require such a design).

That said, it looks like you updated the documentation to add "the iterables are collected immediately rather than lazily;", which, if considered a guarantee, rather than a warning, would make this a breaking change even ignoring the "cancel vs. wait" issue.

Do you have any suggestions? If strict adherence to your newly (as of late 2017) documented behavior is needed, I suppose I could change the default behavior from "reasonable prefetch" to "exhaustive prefetch", so when prefetch isn't passed, every task is submitted, but it would be kind of annoying to lose the "good by default" behavior of limited prefetching.

The reason I cancelled rather than waiting on the result is that I was trying to follow the normal use pattern for map; since the results are yielded lazily, if the iterator goes away or is closed explicitly (or you explicitly shut down the executor), you're done; having the outstanding futures complete when you're not able to see the results means you're either:

1. Expecting the tasks to complete without running out the Executor.map (which doesn't work with Py3's map at all, so the analogy to map should allow it; if you don't run it out, you have no guarantees anything was done)
2. Not planning to use any further results (in which case running any submitted but unscheduled futures means doing work no one can see the results of)

Reply:
Actually, I think there are two problems to discuss:
1) What happens when `shutdown(wait=True)` is called. Currently it waits for all outstanding tasks. I don't think we can change that (the explicit `wait` flag exists for a reason).
2) Whether `map()` can be silently switched to a lazy mode of operation. There's a (perhaps minor) problem with that. Currently, if one of iterables raises an error, `map()` propagates the exception. With your proposal, the exception may be raised later inside the result iterator.

I think 2) might easily be worked around by introducing a separate method (`lazy_map`?).

It seems it would be good to discuss those questions on the [mailing-list](https://mail.python.org/mailman/listinfo/python-dev).

Reply:
Yeah, the problem with using the "lazy_map" name is that it feels like recreating the same annoying distinctions between map and imap from the Py2 era.

If it's necessary to gain acceptance, I could change the default behavior to use prefetch=sys.maxsize - self._max_workers. It would match the pre-existing behavior for just about anything that conceivably worked before (modulo the tiny differences in memory usage of deque vs. list for storing the futures) since:

1. All tasks would be submitted fully up front, so shutdown(wait=True) would in fact wait on them (and no further calls to submit would occur in the generator, so submitting wouldn't occur post-shutdown, which would raise a RuntimeError and cause the cancellation
2. It wouldn't be lazy for anything by default (it would either work eagerly or crash, in the same manner it currently behaves)

If you passed a reasonable prefetch, you wouldn't have these behaviors (and we should document that interaction), but at least existing code would continue to work identically.

Reply:
I don't have a strong opinion. I think discussing those alternatives on the ML, to gather more opinions and arguments, would be useful.